resources/js/components/ui/TranslatedText.tsx

```diff
@@ -0,0 +1,73 @@
+import {Component, JSX, Show, createMemo, mergeProps} from "solid-js";
+import {LangEntryFunc, LangPrefixFunc, getLangEntryFunc} from "../utils";
+import {Capitalize} from "./Capitalize";
+
+interface Props {
+  /** The highest priority source. */
+  override?: () => JSX.Element | undefined;
+  /** The second-highest priority source. */
+  langFunc?: LangEntryFunc | [func?: LangPrefixFunc, subKey?: string];
+  /** Whether to capitalize. Applies to the result of the lang func only. */
+  capitalize?: boolean;
+  /** The last source. */
+  fallbackCode?: string;
+  /* The function that wraps the text, if present. If nothing is displayed then this is not called. */
+  wrapIn?: (text: JSX.Element) => JSX.Element;
+}
+
+/**
+ * Displays a text, using a system of fallbacks. The logic:
+ * - Check props.override:
+ *   - if specified and returns a non-empty value, it is displayed.
+ *   - if specified, but returns an empty value, nothing is displayed.
+ *   - if missing or returns undefined, continue:
+ * - Check props.langFunc:
+ *   - if fully specified and returns a non-empty value, it is displayed, capitalized if props.calitapize.
+ *   - if specified, but returns an empty value, the translation code is displayed
+ *   - if missing, or any part is missing, continue:
+ * - Check props.fallbackCode:
+ *   - if present and non-empty, it is displayed.
+ *   - if missing or empty, nothing is displayed.
+ *
+ * The wrapIn prop allows including a part of HTML only if the value of the TranslatedText is present.
+ * For example this code will only create the <label> element if some text is displayed:
+ *
+ *     <TranslatedText ... wrapIn={text => <label>{text}</label>} />
+ *
+ * This class has some quite complicated logic, which avoids complicated logic in places that show
+ * user text that can originate from multiple sources. An example can be a field label, which can use
+ * a translation system by default, with a possible override for a particular field, and with a fallback
+ * code displayed if nothing else gives a value.
+ */
+export const TranslatedText: Component<Props> = (props) => {
+  const mProps = mergeProps({capitalize: false, wrapIn: (text) => text} satisfies Props, props);
+  const override = createMemo(() => {
+    const value = mProps.override?.();
+    return value === undefined ? undefined : {value, empty: value === "" || (Array.isArray(value) && !value.length)};
```

Comment:
Większość przypadków obsłużona, tylko trochę mnie w oczy kole że jak override zwróci nulla, to zwrócone zostanie ```{ value: null, empty: false }```, traktowałbym nulla jako empty wartość.

Ale tu już się trochę dużo logiki robi. I też nie wiem, może jest case że chcemy nulla printować.

Reply:
No dobra, masz ten null też. Mam nadzieję że to się nie będzie zdarzać.